dmt/A_mappers/ros_A_mapper.py

```diff
@@ -0,0 +1,1024 @@
+        f.write('      offset += sizeof(this->%s);\n' % self.name)
+
+    def convert_from(self, f, spacing):
+        f.write(spacing + '%s = *var;\n' % (self.name))
+
+    def convert_to(self, f, spacing):
+        f.write(spacing + '*var = %s;\n' % (self.name))
+
+
+class MessageDataType(PrimitiveDataType):
+    """ For when our data type is another message. """
+
+    def make_initializer(self, f, trailer):
+        f.write('      %s()%s\n' % (self.name, trailer))
+
+    def serialize(self, f):
+        f.write('      offset += this->%s.serialize(outbuffer + offset);\n' %
+                self.name)
+
+    def deserialize(self, f):
+        f.write('      offset += this->%s.deserialize(inbuffer + offset);\n' %
+                self.name)
+
+    def convert_from(self, f, spacing):
+        f.write(spacing + '%s.fromASN1(&var->%s);\n' % (self.name, self.name))
+
+    def convert_to(self, f, spacing):
+        f.write(spacing + '%s.toASN1(&var->%s);\n' % (self.name, self.name))
+
+
+class StringDataType(PrimitiveDataType):
+    """ Need to convert to signed char *. """
+
+    def __init__(self, name, ty, bytes, string_size=None):
+        self.name = name
+        self.type = ty
+        self.bytes = bytes
+        self.size = string_size
+
+    def make_initializer(self, f, trailer):
+        f.write('      %s("")%s\n' % (self.name, trailer))
+
+    def make_declaration(self, f):
+        # In declaration we need to find the lengh from the type
+        # print ("string length:", find_type_length(self.name))
+
+        # f.write('    typedef const char* _%s_type;\n    _%s_type %s;\n' %
+        # Length should be increased by 1 for NULL termination
+        # length = find_type_length(self.name)
+        f.write('    typedef char _%s_type[%d];\n    _%s_type %s;\n' %
+                (self.name, self.size + 1, self.name, self.name))
+
+    def serialize(self, f):
+        cn = self.name.replace("[", "").replace("]", "")
+        f.write('      uint32_t length_%s = strlen(this->%s);\n' % (cn, self.name))
+        f.write('      varToArr(outbuffer + offset, length_%s);\n' % cn)
+        f.write('      offset += 4;\n')
+        f.write('      memcpy(outbuffer + offset, this->%s, length_%s);\n' %
+                (self.name, cn))
+        f.write('      offset += length_%s;\n' % cn)
+
+    def deserialize(self, f):
+        cn = self.name.replace("[", "").replace("]", "")
+        f.write('      uint32_t length_%s;\n' % cn)
+        f.write('      arrToVar(length_%s, (inbuffer + offset));\n' % cn)
+        f.write('      offset += 4;\n')
+        f.write('      for(unsigned int k= offset; k< offset+length_%s; ++k){\n' %
+                cn)  # shift for null character
+        f.write('        inbuffer[k-1]=inbuffer[k];\n')
+        f.write('      }\n')
+        f.write('      inbuffer[offset+length_%s-1]=0;\n' % cn)
+        f.write('      this->%s = (char *)(inbuffer + offset-1);\n' % self.name)
```

Comment:
Should you be modifying the inbuffer? Shouldn't the bytes be copied?

Instead of adding the variable name to the length local variable, scoping the serialization/deserialization blocks in `{}` would be an alternative.

Reply:
These were directly taken from ROSSerial. I don't think I'm following.

Reply:
Oh, I see, no allocations allowed. That makes sense.

I meant generate the following instead, resolving the `#if`:
```
{
  uint32_t length;
  char *field = inbuffer + offset;
  arrToVar(length, field);
#if this->{self.name} is a pointer
  memmove(field + 3, field + 4, length);
  field += 3;
  field[length] = '\0';
  this->{self.name} = field;
#else  // this->{self.name} is an array
  memcpy(this->{self.name}, field + 4, length);
  this->{self.name}[length] = '\0';
#endif
  offset += length + 4;
}
```